File: ios-native/RelayGo/FlutterBridge.swift
import Foundation
import UIKit
import Flutter
import Combine

/// Bridge between SwiftUI and Flutter engine
/// Communicates via MethodChannel "com.relaygo/bridge"
class FlutterBridge: ObservableObject {
    static let shared = FlutterBridge()

    private var flutterEngine: FlutterEngine?
    private var methodChannel: FlutterMethodChannel?

    // State published to SwiftUI
    @MainActor @Published var isInitialized = false
    @MainActor @Published var isAiReady = false
    @MainActor @Published var isMeshConnected = false
    @MainActor @Published var peerCount = 0
    @MainActor @Published var initProgress = "Starting..."
    @MainActor @Published var isInitializing = false

    // Callbacks for events from Flutter
    var onMeshPacket: ((MeshPacketData) -> Void)?
    var onPeerCountChanged: ((Int) -> Void)?
    var onConnectionStatusChanged: ((String) -> Void)?

    private init() {}

    // MARK: - Engine Setup

    @MainActor
    func startEngine() {
        guard flutterEngine == nil else { return }

        print("[RelayGo] Starting Flutter engine...")

        // Create and run the Flutter engine
        let engine = FlutterEngine(name: "RelayGoEngine")
        engine.run()
        print("[RelayGo] Flutter engine running")

        // Register plugins
        GeneratedPluginRegistrant.register(with: engine)
        print("[RelayGo] Plugins registered")

        flutterEngine = engine

        // Setup method channel
        let channel = FlutterMethodChannel(
            name: "com.relaygo/bridge",
            binaryMessenger: engine.binaryMessenger
        )

        // Handle events pushed from Flutter
        channel.setMethodCallHandler { [weak self] call, result in
            self?.handleFlutterCall(call, result: result)
        }

        methodChannel = channel
        print("[RelayGo] Method channel ready")
    }

    // MARK: - Handle events FROM Flutter

    private func handleFlutterCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        Task { @MainActor in
            switch call.method {
            case "onInitProgress":
                if let args = call.arguments as? [String: Any],
                   let progress = args["progress"] as? String {
                    print("[RelayGo] Progress: \(progress)")
                    self.initProgress = progress
                }
                result(nil)

            case "onMeshPacket":
                if let args = call.arguments as? [String: Any] {
                    let packet = MeshPacketData(from: args)
                    self.onMeshPacket?(packet)
                }
                result(nil)

            case "onPeerCountChanged":
                if let args = call.arguments as? [String: Any],
                   let count = args["count"] as? Int {
                    self.peerCount = count
                    self.onPeerCountChanged?(count)
                }
                result(nil)

            case "onConnectionStatus":
                if let args = call.arguments as? [String: Any],
                   let status = args["status"] as? String {
                    self.isMeshConnected = (status == "connected")
                    self.onConnectionStatusChanged?(status)
                }
                result(nil)

            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }

    // MARK: - Call Flutter methods

    private func invoke(_ method: String, arguments: [String: Any]? = nil) async throws -> [String: Any]? {
        guard let channel = methodChannel else {
            print("[RelayGo] ERROR: Method channel not initialized")
            throw BridgeError.notInitialized
        }

        print("[RelayGo] Invoking Flutter method: \(method)")
        return try await withCheckedThrowingContinuation { continuation in
            DispatchQueue.main.async {
                channel.invokeMethod(method, arguments: arguments) { result in
                    if let error = result as? FlutterError {
                        print("[RelayGo] ERROR from Flutter: \(error.message ?? "unknown")")
                        continuation.resume(throwing: BridgeError.flutterError(error.message ?? "Unknown error"))
                    } else if let dict = result as? [String: Any] {
                        print("[RelayGo] Method \(method) returned successfully")
                        continuation.resume(returning: dict)
                    } else {
                        print("[RelayGo] Method \(method) returned nil")
                        continuation.resume(returning: nil)
                    }
                }
            }
        }
    }

    // MARK: - Public API

    @MainActor
    func initialize() async throws {
        guard !isInitializing else { return }
        isInitializing = true
        initProgress = "Starting Flutter engine..."
        print("[RelayGo] ========== INITIALIZATION STARTED ==========")

        startEngine()

        // Small delay to let engine start
        try await Task.sleep(for: .milliseconds(500))

        initProgress = "Initializing services..."
        print("[RelayGo] Calling Flutter initialize()...")
        let _ = try await invoke("initialize")
        print("[RelayGo] Flutter initialize() completed")

        // Get initial state
        print("[RelayGo] Getting state from Flutter...")
        if let state = try await invoke("getState") {
            isInitialized = state["isInitialized"] as? Bool ?? false
            isAiReady = state["isAiReady"] as? Bool ?? false
            isMeshConnected = state["isMeshConnected"] as? Bool ?? false
            peerCount = state["peerCount"] as? Int ?? 0
            print("[RelayGo] State: initialized=\(isInitialized), aiReady=\(isAiReady), mesh=\(isMeshConnected), peers=\(peerCount)")
        }

        isInitializing = false
        print("[RelayGo] ========== INITIALIZATION COMPLETE ==========")
    }

    @MainActor
    func refreshState() async throws {
        if let state = try await invoke("getState") {
            isInitialized = state["isInitialized"] as? Bool ?? false
            isAiReady = state["isAiReady"] as? Bool ?? false
            isMeshConnected = state["isMeshConnected"] as? Bool ?? false
            peerCount = state["peerCount"] as? Int ?? 0
        }
    }

    // MARK: - AI Methods

    func chat(_ text: String, extractReport: Bool = false) async throws -> ChatResponse {
        let result = try await invoke("chat", arguments: [
            "text": text,
            "extractReport": extractReport
        ])

        guard let result = result else {
            throw BridgeError.invalidResponse
        }

        return ChatResponse(
            text: result["text"] as? String ?? "",
            confidence: result["confidence"] as? String ?? "unverified",
            extraction: result["extraction"] as? [String: Any]
        )
    }

    func transcribe(audioPath: String) async throws -> String {
        let result = try await invoke("transcribe", arguments: [
            "audioPath": audioPath
        ])
        return result?["text"] as? String ?? ""
    }

    func generateAwarenessSummary() async throws -> String {
        let result = try await invoke("generateAwarenessSummary")
        return result?["summary"] as? String ?? "Unable to generate summary"
    }

    // MARK: - Mesh Methods

    @MainActor
    func startMesh() async throws {
        let _ = try await invoke("startMesh")
        isMeshConnected = true
    }

    @MainActor
    func stopMesh() async throws {
        let _ = try await invoke("stopMesh")
        isMeshConnected = false
    }

    func sendSOS(description: String? = nil) async throws {
        var args: [String: Any] = [:]
        if let desc = description {
            args["description"] = desc
        }
        let _ = try await invoke("sendSOS", arguments: args.isEmpty ? nil : args)
    }

    func sendBroadcast(_ message: String) async throws {
        let _ = try await invoke("sendBroadcast", arguments: ["message": message])
    }

    func sendDirectMessage(to peerId: String, message: String) async throws {
        let _ = try await invoke("sendDirectMessage", arguments: [
            "peerId": peerId,
            "message": message
        ])
    }

    func getReports() async throws -> [[String: Any]] {
        let result = try await invoke("getReports")
        return result?["reports"] as? [[String: Any]] ?? []
    }

    func getBroadcasts() async throws -> [[String: Any]] {
        let result = try await invoke("getBroadcasts")
        return result?["broadcasts"] as? [[String: Any]] ?? []
    }

    func getPeers() async throws -> [[String: Any]] {
        let result = try await invoke("getPeers")
        return result?["peers"] as? [[String: Any]] ?? []
    }

    // MARK: - Settings

    func setRelayEnabled(_ enabled: Bool) async throws {
        let _ = try await invoke("setRelayEnabled", arguments: ["enabled": enabled])
    }

    func setDisplayName(_ name: String) async throws {
        let _ = try await invoke("setDisplayName", arguments: ["name": name])
    }
}

// MARK: - Data Types

enum BridgeError: Error, LocalizedError {
    case notInitialized
    case flutterError(String)
    case invalidResponse

    var errorDescription: String? {
        switch self {
        case .notInitialized:
            return "Flutter engine not initialized"
        case .flutterError(let message):
            return "Flutter error: \(message)"
        case .invalidResponse:
            return "Invalid response from Flutter"
        }
    }
}

struct ChatResponse {
    let text: String
    let confidence: String
    let extraction: [String: Any]?

    var isVerified: Bool {
        confidence == "verified"
    }
}

struct MeshPacketData {
    let kind: String // "report" or "message"
    let id: String
    let timestamp: Int
    let source: String
    let body: String?
    let isEmergency: Bool
    let hops: Int

    init(from dict: [String: Any]) {
        kind = dict["kind"] as? String ?? "message"
        id = dict["id"] as? String ?? UUID().uuidString
        timestamp = dict["ts"] as? Int ?? Int(Date().timeIntervalSince1970)
        source = dict["src"] as? String ?? dict["name"] as? String ?? "Unknown"
        body = dict["body"] as? String ?? dict["desc"] as? String
        isEmergency = (dict["urg"] as? Int ?? 0) >= 4 || kind == "report"
        hops = dict["hops"] as? Int ?? 0
    }
}

File: android-native/app/src/main/java/com/relaygo/FlutterBridge.kt
package com.relaygo

import android.content.Context
import io.flutter.embedding.engine.FlutterEngineCache
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/**
 * FlutterBridge manages communication between Kotlin/Compose UI and headless Flutter module.
 * Flutter handles: AI (Cactus), BLE mesh, backend sync
 * Kotlin handles: UI
 */
class FlutterBridge private constructor(context: Context) {

    companion object {
        private const val CHANNEL_NAME = "com.relaygo/bridge"

        @Volatile
        private var instance: FlutterBridge? = null

        fun getInstance(context: Context): FlutterBridge {
            return instance ?: synchronized(this) {
                instance ?: FlutterBridge(context.applicationContext).also { instance = it }
            }
        }
    }

    private val methodChannel: MethodChannel

    // State flows for Compose UI
    private val _isInitialized = MutableStateFlow(false)
    val isInitialized: StateFlow<Boolean> = _isInitialized

    private val _isAiReady = MutableStateFlow(false)
    val isAiReady: StateFlow<Boolean> = _isAiReady

    private val _isMeshConnected = MutableStateFlow(false)
    val isMeshConnected: StateFlow<Boolean> = _isMeshConnected

    private val _peerCount = MutableStateFlow(0)
    val peerCount: StateFlow<Int> = _peerCount

    private val _initProgress = MutableStateFlow("")
    val initProgress: StateFlow<String> = _initProgress

    // Callbacks
    var onMeshPacket: ((Map<String, Any?>) -> Unit)? = null
    var onPeerCountChanged: ((Int) -> Unit)? = null

    init {
        val engine = FlutterEngineCache.getInstance().get(RelayGoApp.FLUTTER_ENGINE_ID)
            ?: throw IllegalStateException("Flutter engine not initialized")

        methodChannel = MethodChannel(engine.dartExecutor.binaryMessenger, CHANNEL_NAME)

        // Handle calls FROM Flutter
        methodChannel.setMethodCallHandler { call, result ->
            when (call.method) {
                "onInitProgress" -> {
                    val progress = call.argument<String>("progress") ?: ""
                    _initProgress.value = progress
                    result.success(null)
                }
                "onMeshPacket" -> {
                    @Suppress("UNCHECKED_CAST")
                    val packet = call.arguments as? Map<String, Any?>
                    packet?.let { onMeshPacket?.invoke(it) }
                    result.success(null)
                }
                "onPeerCountChanged" -> {
                    val count = call.argument<Int>("count") ?: 0
                    _peerCount.value = count
                    onPeerCountChanged?.invoke(count)
                    result.success(null)
                }
                "onConnectionStatus" -> {
                    val status = call.argument<String>("status") ?: ""
                    _isMeshConnected.value = (status == "connected")
                    result.success(null)
                }
                else -> result.notImplemented()
            }
        }
    }

    // ============================================================
    // INITIALIZATION
    // ============================================================

    suspend fun initialize(): Boolean = invokeMethod("initialize").let {
        val success = it["success"] as? Boolean ?: false
        if (success) {
            _isInitialized.value = true
            refreshState()
        }
        success
    }

    // ============================================================
    // AI METHODS
    // ============================================================

    suspend fun transcribe(audioPath: String): String {
        val result = invokeMethod("transcribe", mapOf("audioPath" to audioPath))
        return result["text"] as? String ?: ""
    }

    suspend fun chat(text: String, extractReport: Boolean = false): ChatResponse {
        val result = invokeMethod("chat", mapOf(
            "text" to text,
            "extractReport" to extractReport
        ))
        return ChatResponse(
            text = result["text"] as? String ?: "",
            confidence = result["confidence"] as? String ?: "unverified",
            extraction = (result["extraction"] as? Map<*, *>)?.let { ExtractionData.fromMap(it) }
        )
    }

    suspend fun generateAwarenessSummary(): String {
        val result = invokeMethod("generateAwarenessSummary")
        return result["summary"] as? String ?: ""
    }

    // ============================================================
    // MESH METHODS
    // ============================================================

    suspend fun startMesh() {
        invokeMethod("startMesh")
        _isMeshConnected.value = true
    }

    suspend fun stopMesh() {
        invokeMethod("stopMesh")
        _isMeshConnected.value = false
    }

    suspend fun sendSOS(description: String? = null) {
        val args = mutableMapOf<String, Any>()
        description?.let { args["description"] = it }
        invokeMethod("sendSOS", args)
    }

    suspend fun sendBroadcast(message: String) {
        invokeMethod("sendBroadcast", mapOf("message" to message))
    }

    suspend fun sendDirectMessage(peerId: String, message: String) {
        invokeMethod("sendDirectMessage", mapOf(
            "peerId" to peerId,
            "message" to message
        ))
    }

    @Suppress("UNCHECKED_CAST")
    suspend fun getReports(): List<Map<String, Any?>> {
        val result = invokeMethod("getReports")
        return result["reports"] as? List<Map<String, Any?>> ?: emptyList()
    }

    @Suppress("UNCHECKED_CAST")
    suspend fun getBroadcasts(): List<Map<String, Any?>> {
        val result = invokeMethod("getBroadcasts")
        return result["broadcasts"] as? List<Map<String, Any?>> ?: emptyList()
    }

    @Suppress("UNCHECKED_CAST")
    suspend fun getPeers(): List<Map<String, Any?>> {
        val result = invokeMethod("getPeers")
        return result["peers"] as? List<Map<String, Any?>> ?: emptyList()
    }

    // ============================================================
    // SETTINGS
    // ============================================================

    suspend fun setRelayEnabled(enabled: Boolean) {
        invokeMethod("setRelayEnabled", mapOf("enabled" to enabled))
    }

    suspend fun setDisplayName(name: String) {
        invokeMethod("setDisplayName", mapOf("name" to name))
    }

    private suspend fun refreshState() {
        val state = invokeMethod("getState")
        _isInitialized.value = state["isInitialized"] as? Boolean ?: false
        _isAiReady.value = state["isAiReady"] as? Boolean ?: false
        _isMeshConnected.value = state["isMeshConnected"] as? Boolean ?: false
        _peerCount.value = state["peerCount"] as? Int ?: 0
    }

    // ============================================================
    // PRIVATE HELPERS
    // ============================================================

    @Suppress("UNCHECKED_CAST")
    private suspend fun invokeMethod(
        method: String,
        arguments: Map<String, Any?> = emptyMap()
    ): Map<String, Any?> = suspendCancellableCoroutine { continuation ->
        methodChannel.invokeMethod(method, arguments, object : MethodChannel.Result {
            override fun success(result: Any?) {
                continuation.resume(result as? Map<String, Any?> ?: emptyMap())
            }

            override fun error(code: String, message: String?, details: Any?) {
                continuation.resumeWithException(BridgeException("$code: $message"))
            }

            override fun notImplemented() {
                continuation.resumeWithException(BridgeException("Method $method not implemented"))
            }
        })
    }
}

class BridgeException(message: String) : Exception(message)

data class ChatResponse(
    val text: String,
    val confidence: String,
    val extraction: ExtractionData?
)

data class ExtractionData(
    val type: String,
    val urgency: Int,
    val hazards: List<String>,
    val description: String
) {
    companion object {
        @Suppress("UNCHECKED_CAST")
        fun fromMap(map: Map<*, *>): ExtractionData {
            return ExtractionData(
                type = map["type"] as? String ?: "other",
                urgency = (map["urgency"] as? Number)?.toInt() ?: 3,
                hazards = map["hazards"] as? List<String> ?: emptyList(),
                description = map["description"] as? String ?: ""
            )
        }
    }
}
